---
title: "Landscape Genetics"
author: "Connor French"
date: "12/22/2020"
output: 
  html_document:
    theme: flatly
    highlight: tango
    toc: true
    toc_float: true
    collapsed: false
    code_folding: show
---


```{r setup, message=FALSE}
library(here)
library(adegenet)
library(vcfR)
library(gdistance)
library(raster)
library(rnaturalearth)
library(spatialEco)
library(PopGenome)
library(BBmisc)
library(patchwork)
library(tidyverse)
library(radish)
library(sf)

# source Wang 2013's MMRR script. I obtained it from here: https://nature.berkeley.edu/wanglab/data/
# I modified it to return an AICc score and loglikelihood for model comparison  
source(here("R", "MMRR_function.R"))
```


Read in the localities, land cover and projected SDM models for generating friction surfaces. The land cover was processed in the `process_land_cover.Rmd` report. The projections were made in the `species_distribution_modeling.Rmd`
report. Not going to use LGM since it's almost completely correlated with current climate.  

```{r}
# localities
occs <- read_csv(here("data", "sample_latlongs.csv")) %>% 
  filter(!duplicated(longitude)) %>% 
  st_as_sf(coords = c("longitude", "latitude"),
           crs = 4326) %>% 
  st_geometry() %>% 
  as("Spatial")

# forest cover
# 0 = forested, 1 = non-forested
fc <- raster(here("output", "rasters", "forest_cover.tif"))

# current prediction
pred <- raster(here("output", "rasters", "current_projection.tif")) %>% 
  crop(fc) %>% 
  mask(fc)

# lgm prediction
lgm_pred <- raster(here("output", "rasters", "lgm_projection.tif")) %>% 
  crop(pred) %>% 
  mask(pred) %>% 
  crop(fc) %>% 
  mask(fc)

# current has a few cells on the edge that are NA 
fc <- fc %>% 
  crop(pred) %>% 
  mask(pred) 

# forest friction, 1:10 ratio
fc_1_10 <- fc
values(fc_1_10)[values(fc_1_10) == 1] <- 10
values(fc_1_10)[values(fc_1_10) == 0] <- 1

# forest friction, 1:100 ratio
fc_1_100 <- fc
values(fc_1_100)[values(fc_1_100) == 1] <- 100
values(fc_1_100)[values(fc_1_100) == 0] <- 1

# forest friction, 1:1000 ratio
fc_1_1000 <- fc
values(fc_1_1000)[values(fc_1_1000) == 1] <- 1000
values(fc_1_1000)[values(fc_1_1000) == 0] <- 1

# forest friction, 1:10000 ratio
fc_1_10000 <- fc
values(fc_1_10000)[values(fc_1_10000) == 1] <- 10000
values(fc_1_10000)[values(fc_1_10000) == 0] <- 1

#check out correlation between current and lgm climate
cormat <- suppressWarnings(raster::corLocal(pred, lgm_pred)) %>% 
  as.matrix()

avg_cor <- mean(cormat[!is.na(cormat)])
paste("Average correlation between current and lgm climate is", round(avg_cor, 2))
```


# Least-cost paths approach

Generate friction surface for landscape genetics. Given that we don't know much about how geckos traverse the landscape and the landscape is broken into large cell sizes relative to an individual gecko's dispersal potential, I'm going to first take a least-cost paths approach, then try out the commute distance. 

```{r cache=TRUE}
#read in all localities where there is genetic sampling
gen_locs_path <- here("data", "sample_latlongs.csv")
gen_locs <-
  read_csv(gen_locs_path) %>% 
  dplyr::select(longitude, latitude) %>% 
  filter(!duplicated(longitude)) # single observation per locality for distance calculations

coordinates(gen_locs) <- ~ longitude + latitude

#geographic distances for pure IBD hypothesis
geo_dist <- raster::pointDistance(gen_locs, lonlat = TRUE) %>% 
  as.dist() %>% 
  as.matrix()

###transition matrices with geographic corrections
tr_current_lcp <- pred %>%
  spatialEco::raster.invert() %>%  #invert the SDM values
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix #correct for geographic distortion
  gdistance::geoCorrection("c") #correct for geographic distortion

tr_fc_1_10_lcp <- fc_1_10 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("c") #correct for geographic distortion

tr_fc_1_100_lcp <- fc_1_100 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("c") #correct for geographic distortion distortion

tr_fc_1_1000_lcp <- fc_1_1000 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("c") #correct for geographic distortion

tr_fc_1_10000_lcp <- fc_1_10000 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("c") #correct for geographic distortion

```

```{r cache=TRUE}
# normalize the distances for accurate comparison in regression
lcp_current <- 
  costDistance(tr_current_lcp, coordinates(gen_locs)) %>% 
  as.matrix()

lcp_fc_1_10 <- 
  costDistance(tr_fc_1_10_lcp, coordinates(gen_locs)) %>% 
  as.matrix()

lcp_fc_1_100 <- 
  costDistance(tr_fc_1_100_lcp, coordinates(gen_locs)) %>% 
  as.matrix()

lcp_fc_1_1000 <- 
  costDistance(tr_fc_1_1000_lcp, coordinates(gen_locs)) %>% 
  as.matrix()

lcp_fc_1_10000 <- 
  costDistance(tr_fc_1_10000_lcp, coordinates(gen_locs)) %>% 
  as.matrix()
```

Write distance matrices to file  

```{r, eval=FALSE}


```


Calculate a pairwise Dxy matrix for our genetic distance matrix. 

```{r eval=FALSE}
pops <- read_csv(here("data", "pops_file.csv")) %>% 
  mutate(
    locality = case_when(
      near(longitude, -105.0444, tol = 0.0001) ~ "ST",
      near(longitude, -105.2503, tol = 0.0001) ~ "PC",
      near(longitude, -105.0980, tol = 0.0001) ~ "LG",
      near(longitude, -104.8789, tol = 0.0001) ~ "P-35",
      near(longitude, -104.8027, tol = 0.0001) ~ "PZ",
      near(longitude, -104.5181, tol = 0.0001) ~ "CH",
      near(longitude, -104.7663, tol = 0.0001) ~ "PT",
      near(longitude, -104.9325, tol = 0.0001) ~ "LS",
      near(longitude, -104.8500, tol = 0.0001) ~ "P-30"
    ) %>% as.factor()
  )

pops_list <- pops %>% 
  select(locality, individuals) %>% 
  group_split(locality) %>% 
  map(pull, individuals)

# move the 70% missing vcf to a folder for popgenome if it's not already there (PopGenome needs a unique folder to work out of)
if (!dir.exists(here("data", "popgenome"))) {
  dir.create(here("data", "popgenome"))
  file.move(here("data", "missing_70_thin.recode.vcf"),
            here("data", "popgenome", "missing_70_thin.recode.vcf"))
}

gen_data <- PopGenome::readData(here("data", "popgenome"),
                                populations = pops_list,
                                format = "VCF",
                                include.unknown = TRUE)



# # calc pairwise nucleotide divergence
# gen_data <- PopGenome::diversity.stats.between(object = gen_data, 
#                                               nucleotide.mode = TRUE)
# gen_data <- PopGenome::F_ST.stats(object = gen_data)
# 
# fst_vec <- gen_data@nuc.F_ST.pairwise[,1]
# fst_vec <- fst_vec/(1 - fst_vec)
# 
# fst_mat <- matrix(nrow = 9, ncol = 9)
# 
# fst_mat[upper.tri(fst_mat)] <- fst_vec
# fst_mat[lower.tri(fst_mat)] <- fst_vec
# 
# 
# 
# # write to file
# #write.table(dxy_mat,
# #            file = here("output", "spreadsheets", "dxy_matrix.txt"))
```



Since strong correlation among predictors will violate the assumptions of a regression, I'll investigate correlations among the predictor variables. I'm using the MMRR to account for the effect of pairwise comparisons. MMRR standardizes predictors on its own, so I don't need to do that myself.

First, let's read in the data if I'm picking this back up again.
```{r}

# list of predictors
lcp_current <-
  read.table(file = here("output", "spreadsheets", "lcp_dist_current.txt")) %>%
  as.matrix() 

lcp_lgm <-
  read.table(file = here("output", "spreadsheets", "lcp_dist_lgm.txt")) %>%
  as.matrix() 

geo_dist <-
  read.table(file = here("output", "spreadsheets", "geo_distance_matrix.txt")) %>%
  as.matrix() 

# to remove the necessity of running the previous chunk
dxy_mat <- 
  read.table(file = here("output", "spreadsheets", "dxy_matrix.txt")) %>%
  as.matrix() 


```



Going forward, I'm going to only include current climate, forest cover, and geographic distance since the present and LGM climate are strongly correlated. We won't be able to really tell which variable is explaining the variation in genetic distance.

## Current + FC 1:10 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_current, lcp_fc_1_10, geo_dist)

set.seed(2344)
mmrr_obj_1_10 <- MMRR(fst_mat, pred_vars, nperm = 1000)


mmrr_obj_1_10
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_10, file = here("output", "mmrr_lcp_1_10.rds"))
```

## Current + FC 1:100 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_current, lcp_fc_1_100, geo_dist)

set.seed(2323444)
mmrr_obj_1_100 <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_100
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_100, file = here("output", "mmrr_lcp_1_100.rds"))
```

## Current + FC 1:1000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_current, lcp_fc_1_1000, geo_dist)

set.seed(323444)
mmrr_obj_1_1000 <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_1000
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_1000, file = here("output", "mmrr_lcp_1_1000.rds"))
```


## Current + FC 1:10000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_current, lcp_fc_1_10000, geo_dist)

set.seed(323344)
mmrr_obj_1_10000 <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_10000
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_10000, file = here("output", "mmrr_lcp_1_10000.rds"))
```

## Current + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_current, geo_dist)

set.seed(3444999)
mmrr_obj_current<- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_current
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_current, file = here("output", "mmrr_lcp_current.rds"))
```

## FC 1:10 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_fc_1_10, geo_dist)

set.seed(34441283)
mmrr_obj_1_10_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_10_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_10_geo, file = here("output", "mmrr_lcp_1_10_geo.rds"))
```

## FC 1:100 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_fc_1_100, geo_dist)

set.seed(355544)
mmrr_obj_1_100_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_100_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_100_geo, file = here("output", "mmrr_lcp_1_100_geo.rds"))
```

## FC 1:1000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_fc_1_1000, geo_dist)

set.seed(55544)
mmrr_obj_1_1000_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_1000_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_1000_geo, file = here("output", "mmrr_lcp_1_1000_geo.rds"))
```

## FC 1:10000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(lcp_fc_1_10000, geo_dist)

set.seed(3394)
mmrr_obj_1_10000_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_1_10000_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_1_10000_geo, file = here("output", "mmrr_lcp_1_10000_geo.rds"))
```


## Geo Dist

```{r cache=TRUE}
pred_vars <- list(geo_dist)

set.seed(3449001)
mmrr_obj_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_geo, file = here("output", "mmrr_lcp_geo.rds"))
```

## Statistic table

```{r}
get_mmrr_stats <- function(mmrr_obj) {
  rsq <- mmrr_obj$r.squared
  fstat <- mmrr_obj$Fstatistic
  fp <- mmrr_obj$Fpvalue
  aicc <- mmrr_obj$aicc
  log_lik <- mmrr_obj$log_lik
  
  return(
    tibble(
      rsquared = rsq,
      fstatistic = fstat,
      fpvalue = fp,
      aicc = aicc,
      log_lik = log_lik
    )
  )
}

lcp_mmrr_objects <- list(
  lcp_1_10_curr_geo = mmrr_obj_1_10,
  lcp_1_100_curr_geo = mmrr_obj_1_100,
  lcp_1_1000_curr_geo = mmrr_obj_1_1000,
  lcp_1_10000_curr_geo = mmrr_obj_1_10000,
  lcp_1_10_geo = mmrr_obj_1_10_geo,
  lcp_1_100_geo = mmrr_obj_1_100_geo,
  lcp_1_1000_geo = mmrr_obj_1_1000_geo,
  lcp_1_10000_geo = mmrr_obj_1_10000_geo,
  lcp_curr_geo = mmrr_obj_current,
  lcp_geo = mmrr_obj_geo
)

mmrr_stats_lcp <- map_dfr(lcp_mmrr_objects, get_mmrr_stats) %>% 
  mutate(hypothesis = names(lcp_mmrr_objects),
         diff_aicc = aicc - min(aicc))

mmrr_stats_lcp %>% 
  arrange(aicc)
```


## LCP scatterplot
```{r}
mmrr_df <- tibble(
  geo_dist =  geo_dist[lower.tri(geo_dist)],
  lcp_dist = lcp_fc_1_10[lower.tri(lcp_fc_1_10)],
  fst = fst_mat[lower.tri(fst_mat)]
) %>% 
  mutate(mmrr_model = mmrr_obj_1_10_geo$coefficients[2] * lcp_dist + 
           mmrr_obj_1_10_geo$coefficients[3] * geo_dist)
```


```{r}
# current_plot <- ggplot(mmrr_df, aes(x = lcp_dist, y = fst)) +
#   geom_point() +
#   geom_smooth(method = lm, se = FALSE, color = "black") +
#   labs(x = "Forest Cover", y = "Fst")
# 
# geo_plot <- ggplot(mmrr_df, aes(x = geo_dist, y = fst)) +
#   geom_point() +
#   #geom_smooth(method = lm, se = FALSE, color = "black") +
#   labs(x = "Geographic Distance", y = "Fst")

full_model_plot <- ggplot(mmrr_df, aes(x = mmrr_model, y = fst)) +
  geom_point() +
  geom_smooth(method = lm, se = FALSE, color = "black") +
  labs(x = paste0("(", round(mmrr_obj_1_10_geo$coefficients[2], 3), 
  " * Forest Cover) + (", 
  round(mmrr_obj_1_10_geo$coefficients[3], 3),
  " * Geo Distance)"), y = "Fst")


full_model_plot
```

# Commute distance approach

Create a resistance distance matrix using the commute distance approach. Since LGM and current climate are highly correlated, we're only considering current climate and forest cover  

Transition matrices with appropriate geographic correction for random walk  
```{r}
###transition matrices with geographic corrections
tr_current_commute <- pred %>%
  spatialEco::raster.invert() %>%  #invert the SDM values
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix #correct for geographic distortion
  gdistance::geoCorrection("r", scl = TRUE) #correct for geographic distortion

tr_fc_1_10_commute <- fc_1_10 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("r", scl = TRUE) #correct for geographic distortion

tr_fc_1_100_commute <- fc_1_100 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("r", scl = TRUE) #correct for geographic distortion

tr_fc_1_1000_commute <- fc_1_1000 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  ggdistance::geoCorrection("r", scl = TRUE) #correct for geographic distortion

tr_fc_1_10000_commute <- fc_1_10000 %>%
  gdistance::transition(mean, directions = 8) %>%  #create a transition matrix
  gdistance::geoCorrection("r", scl = TRUE) #correct for geographic distortion

```


Commute distances  

```{r}
commute_current <- 
  commuteDistance(tr_current_commute, coordinates(gen_locs)) %>% 
  as.matrix()

commute_fc_1_10 <- 
  commuteDistance(tr_fc_1_10_commute, coordinates(gen_locs)) %>% 
  as.matrix()

commute_fc_1_100 <- 
  commuteDistance(tr_fc_1_100_commute, coordinates(gen_locs)) %>% 
  as.matrix()

commute_fc_1_1000 <- 
  commuteDistance(tr_fc_1_1000_commute, coordinates(gen_locs)) %>% 
  as.matrix()

commute_fc_1_10000 <- 
  commuteDistance(tr_fc_1_10000_commute, coordinates(gen_locs)) %>% 
  as.matrix()

```

```{r, eval=FALSE}
### Write commute distance matrices to file
write.table(commute_current,
            file = here("output", "spreadsheets", "commute_distance_matrix.txt"))

write.table(commute_fc_1_10,
            file = here("output", "spreadsheets", "commute_fc_1_10_distance_matrix.txt"))

write.table(commute_fc_1_100,
            file = here("output", "spreadsheets", "commute_fc_1_100_distance_matrix.txt"))

write.table(commute_fc_1_1000,
            file = here("output", "spreadsheets", "commute_fc_1_1000_distance_matrix.txt"))

write.table(commute_fc_1_10000,
            file = here("output", "spreadsheets", "commute_fc_1_10000_distance_matrix.txt"))
```

## Current + FC 1:10 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_current, commute_fc_1_10, geo_dist)

set.seed(238497)
mmrr_obj_commute_1_10 <- MMRR(fst_mat, pred_vars, nperm = 1000)


mmrr_obj_commute_1_10
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_10, file = here("output", "mmrr_commute_1_10.rds"))
```

## Current + FC 1:100 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_current, commute_fc_1_100, geo_dist)

set.seed(11123)
mmrr_obj_commute_1_100 <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_100
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_100, file = here("output", "mmrr_commute_1_100.rds"))
```

## Current + FC 1:1000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_current, commute_fc_1_1000, geo_dist)

set.seed(1874)
mmrr_obj_commute_1_1000 <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_1000
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_1000, file = here("output", "mmrr_commute_1_1000.rds"))
```


## Current + FC 1:10000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_current, commute_fc_1_10000, geo_dist)

set.seed(31111144)
mmrr_obj_commute_1_10000 <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_10000
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_10000, file = here("output", "mmrr_commute_1_10000.rds"))
```

## Current + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_current, geo_dist)

set.seed(444919)
mmrr_obj_commute_current<- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_current
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_current, file = here("output", "mmrr_commute_current.rds"))
```

## FC 1:10 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_fc_1_10, geo_dist)

set.seed(3441111)
mmrr_obj_commute_1_10_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_10_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_10_geo, file = here("output", "mmrr_commute_1_10_geo.rds"))
```

## FC 1:100 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_fc_1_100, geo_dist)

set.seed(1544)
mmrr_obj_commute_1_100_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_100_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_100_geo, file = here("output", "mmrr_commute_1_100_geo.rds"))
```

## FC 1:1000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_fc_1_1000, geo_dist)

set.seed(5111)
mmrr_obj_commute_1_1000_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_1000_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_1000_geo, file = here("output", "mmrr_commute_1_1000_geo.rds"))
```

## FC 1:10000 + Geo Dist

```{r cache=TRUE}
pred_vars <- list(commute_fc_1_10000, geo_dist)

set.seed(3394111)
mmrr_obj_commute_1_10000_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_1_10000_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_1_10000_geo, file = here("output", "mmrr_commute_1_10000_geo.rds"))
```


## Geo Dist

```{r cache=TRUE}
pred_vars <- list(geo_dist)

set.seed(3441111)
mmrr_obj_commute_geo <- MMRR(fst_mat, pred_vars, nperm = 1000)

mmrr_obj_commute_geo
```

```{r, eval=FALSE}
# save mmrr to file
saveRDS(mmrr_obj_commute_geo, file = here("output", "mmrr_commute_geo.rds"))
```

## Statistic table

```{r}
get_mmrr_stats <- function(mmrr_obj) {
  rsq <- mmrr_obj$r.squared
  fstat <- mmrr_obj$Fstatistic
  fp <- mmrr_obj$Fpvalue
  aicc <- mmrr_obj$aicc
  
  return(
    tibble(
      rsquared = rsq,
      fstatistic = fstat,
      fpvalue = fp,
      aicc = aicc
    )
  )
}

commute_mmrr_objects <- list(
  commute_1_10_curr_geo = mmrr_obj_commute_1_10,
  commute_1_100_curr_geo = mmrr_obj_commute_1_100,
  commute_1_1000_curr_geo = mmrr_obj_commute_1_1000,
  commute_1_10000_curr_geo = mmrr_obj_commute_1_10000,
  commute_1_10_geo = mmrr_obj_commute_1_10_geo,
  commute_1_100_geo = mmrr_obj_commute_1_100_geo,
  commute_1_1000_geo = mmrr_obj_commute_1_1000_geo,
  commute_1_10000_geo = mmrr_obj_commute_1_10000_geo,
  commute_curr_geo = mmrr_obj_commute_current,
  commute_geo = mmrr_obj_commute_geo
)

mmrr_stats_commute <- map_dfr(commute_mmrr_objects, get_mmrr_stats) %>% 
  mutate(hypothesis = names(commute_mmrr_objects),
         diff_aicc = aicc - min(aicc))

mmrr_stats_commute %>% 
  arrange(diff_aicc)


mmrr_aic_list <- commute_mmrr_objects <- list(
  mmrr_obj_commute_1_10$fit,
  mmrr_obj_commute_1_100$fit,
  mmrr_obj_commute_1_1000$fit,
  mmrr_obj_commute_1_10000$fit,
  mmrr_obj_commute_1_10_geo$fit,
  mmrr_obj_commute_1_100_geo$fit,
  mmrr_obj_commute_1_1000_geo$fit,
  mmrr_obj_commute_1_10000_geo$fit,
  mmrr_obj_commute_current$fit,
  mmrr_obj_commute_geo$fit
)

mmrr_aic_table <- AICcmodavg::aictab(cand.set = mmrr_aic_list, 
                                     modnames = names(commute_mmrr_objects))

mmrr_aic_table
```

Write table to file  
```{r, eval=FALSE}
write_csv(mmrr_aic_table, here("output", "spreadsheets", "mmrr_aic_table.csv"))
```


Data frame for the commute distance MMRR
```{r}
mmrr_df_commute <- tibble(
  geo_dist =  unfold(geo_dist)[,1],
  commute_dist = unfold(commute_fc_1_10000)[,1],
  fst = unfold(fst_mat)[,1]
) %>% 
  mutate(mmrr_model = mmrr_obj_commute_1_10000_geo$coefficients[2] * commute_dist + 
           mmrr_obj_commute_1_10000_geo$coefficients[3] * geo_dist)
```

```{r}
full_model_plot <- ggplot(mmrr_df_commute, aes(x = mmrr_model, y = fst)) +
  geom_point() +
  geom_smooth(method = lm, se = FALSE, color = "black") +
  labs(x = paste0("(", round(mmrr_obj_commute_1_10000_geo$coefficients[2], 3), 
  " * Forest Cover) + (", 
  round(mmrr_obj_commute_1_10000_geo$coefficients[3], 3),
  " * Geo Distance)"), y = "Fst") +
  theme_bw()

full_model_plot
```




# Generate resistance surfaces

Read in genetic distances. This Weir and Cockram's (1984) FST matrix was estimated using the `diveRsity` R package (v1.9.90) function `bigDivPart`.   
```{r}
fst_mat <- 
  read.table(file = here("output", "spreadsheets", "fst_matrix.txt")) %>%
  as.matrix() 

# linearize
fst_mat <- fst_mat / (1-fst_mat)

```


I'm using the `radish` package to convert my covariates to resistance surfaces for modeling.  

```{r}
# scaling spatial covariates helps avoid numeric overflow
covariates <- raster::stack(list(current = raster::scale(pred),
                                 #lgm = raster::scale(lgm_pred),
                                 forest_cover = fc_cat))

surface <- conductance_surface(covariates, occs, directions = 8)

```

### Fit models

```{r}
set.seed(77789)
fit_additive <- radish(fst_mat ~ current + forest_cover, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(1123)
fit_interaction <- radish(fst_mat ~ current * forest_cover, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(99890)
fit_current <- radish(fst_mat ~ current, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(116247)
fit_lc <- radish(fst_mat ~ forest_cover, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(192387)
fit_ibd <- radish(fst_mat ~ 1, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

```


Summaries of each model.  

```{r}
tidy_fit <- function(r_fit) {
  tib_fit <- tibble(
    aic = r_fit$aic,
    loglik = r_fit$loglik,
    df = fit_additive$df
  )
  
  return(tib_fit)
}


fits <- list(
  fit_additive = fit_additive,
  fit_interaction = fit_interaction,
  fit_current = fit_current,
  fit_lc = fit_lc,
  fit_ibd = fit_ibd
)

fit_df <- map_dfr(fits, tidy_fit) %>% 
  mutate(fit_name = names(fits))

summary(fit_additive)
summary(fit_interaction)
summary(fit_current)
summary(fit_lc)
summary(fit_ibd)
```


Likelihood ratio tests measuring each model's fit compared to IBD.  

```{r}
anova(fit_additive, fit_ibd)
anova(fit_interaction, fit_ibd)
anova(fit_current, fit_ibd)
anova(fit_lc, fit_ibd)
```

### Visualize fit

This is the fit for the additive model.  

```{r}
fit_mat_additive <- fitted(fit_additive, "distance")
fit_mat_additive <- fit_mat_additive[upper.tri(fit_mat_additive)]

fst_mat_additive <- fst_mat[upper.tri(fst_mat)]



plot(fitted(fit_lc, "distance"), fst_mat, pch = 19,
     xlab = "Optimized resistance distance", ylab = "Fst")


```


### Visualize conductance surface

```{r}
# visualise estimated conductance surface and asymptotic confidence intervals
fitted_conductance <- conductance(surface, fit_additive, quantile = 0.95)

plot(fitted_conductance[["est"]], 
     main = "Fitted conductance surface\n(Current Climate + Forest Cover)")
plot(occs, add = TRUE)
```

### IBD visualization

```{r}
# visualise likelihood surface across grid (takes awhile)
theta <- as.matrix(expand.grid(current = seq(-1, 1, length.out=21), 
                               forest_cover = seq(-1, 1, length.out=21)))

# calculate resistance distances across grid
distances <- radish_distance(theta, ~current + forest_cover, 
                             surface, radish::loglinear_conductance)


# index for the matrix that contains IBD (where theta for both dimensions = 0)
ibd <- which(theta[,1] == 0 & theta[,2] == 0)

plot(distances$distance[,,ibd], fst_mat, pch = 19, 
     xlab = "Null resistance distance (IBD)", ylab = "Fst")
```


# No CH Resistance surfaces

Read in genetic distances. This Weir and Cockram's (1984) FST matrix was estimated using the `diveRsity` R package (v1.9.90) function `bigDivPart`.   
```{r}
fst_mat <- 
  read.table(file = here("output", "spreadsheets", "fst_matrix.txt")) %>%
  as.matrix() 

fst_no9 <- fst_mat[rownames(fst_mat) != "pop_9", colnames(fst_mat) != "pop_9"]

# linearize
fst_no9 <- fst_no9 / (1-fst_no9)

```


I'm using the `radish` package to convert my covariates to resistance surfaces for modeling.  

```{r}
occs_no9 <- pops %>% 
  filter(!duplicated(longitude),
         pops != 9) %>% 
  st_as_sf(coords = c("longitude", "latitude"),
           crs = 4326) %>% 
  st_geometry() %>% 
  as("Spatial")

# scaling spatial covariates helps avoid numeric overflow
covariates <- raster::stack(list(current = raster::scale(pred),
                                 #lgm = raster::scale(lgm_pred),
                                 forest_cover = fc_cat))

surface <- conductance_surface(covariates, occs_no9, directions = 8)

```

### Fit models

```{r}
set.seed(819)
fit_additive <- radish(fst_no9 ~ current + forest_cover, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(231)
fit_interaction <- radish(fst_no9 ~ current * forest_cover, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(9910)
fit_current <- radish(fst_no9 ~ current, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(1247)
fit_lc <- radish(fst_no9 ~ forest_cover, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

set.seed(3187)
fit_ibd <- radish(fst_no9 ~ 1, 
                   data = surface, 
                   conductance_model = radish::loglinear_conductance, 
                   measurement_model = radish::mlpe)

```


Summaries of each model.  

```{r}
summary(fit_additive)
summary(fit_interaction)
summary(fit_current)
summary(fit_lc)
summary(fit_ibd)



fits <- list(
  fit_additive = fit_additive,
  fit_interaction = fit_interaction,
  fit_current = fit_current,
  fit_lc = fit_lc,
  fit_ibd = fit_ibd
)

fit_df <- map_dfr(fits, tidy_fit) %>% 
  mutate(fit_name = names(fits))
```


Likelihood ratio tests measuring each model's fit compared to IBD.  

```{r}
anova(fit_additive, fit_ibd)
anova(fit_interaction, fit_ibd)
anova(fit_current, fit_ibd)
anova(fit_lc, fit_ibd)
anova(fit_interaction, fit_lc)
```

### Visualize fit

This is the fit for the additive model.  

```{r}
plot(fitted(fit_interaction, "distance"), fst_no9, pch = 19,
     xlab = "Optimized resistance distance", ylab = "Fst / (1 - Fst)")
```


### Visualize conductance surface

```{r}
# visualise estimated conductance surface and asymptotic confidence intervals
fitted_conductance <- conductance(surface, fit_interaction, quantile = 0.95)

est_conductance <- fitted_conductance[["est"]]

# some values at the edge of the landscape are unreasonably large, so making those NA values
# they don't impact migration between populations
values(est_conductance)[values(est_conductance) > 10000] <- NA


plot(log10(est_conductance), 
     main = "Fitted conductance surface\n(Current Climate * Forest Cover)",
     )
plot(occs_no9, add = TRUE)
```

Crop to narrower area  

```{r}
buffer_no9 <- st_as_sf(occs_no9) %>% 
  st_union() %>% 
  st_convex_hull() %>% 
  st_buffer(0.1)

plot(buffer_no9)
```


### IBD visualization

```{r}
# visualise likelihood surface across grid (takes awhile)
theta <- as.matrix(expand.grid(current = seq(-1, 1, length.out=21), 
                               forest_cover = seq(-1, 1, length.out=21)))

# calculate resistance distances across grid
distances <- radish_distance(theta, ~current + forest_cover, 
                             surface, radish::loglinear_conductance)


# index for the matrix that contains IBD (where theta for both dimensions = 0)
ibd <- which(theta[,1] == 0 & theta[,2] == 0)

ibd_mat <- distances$distance[,,ibd]

plot(ibd_mat, fst_no9, pch = 19, 
     xlab = "Null resistance distance (IBD)", ylab = "Fst / (1- Fst)")
```

# Visualize missingness

Read in data  

```{r}
pops <- read_csv(here("data", "pops_file.csv")) %>% 
  mutate(
    locality = case_when(
      near(longitude, -105.0444, tol = 0.0001) ~ "ST",
      near(longitude, -105.2503, tol = 0.0001) ~ "PC",
      near(longitude, -105.0980, tol = 0.0001) ~ "LG",
      near(longitude, -104.8789, tol = 0.0001) ~ "P-35",
      near(longitude, -104.8027, tol = 0.0001) ~ "PZ",
      near(longitude, -104.5181, tol = 0.0001) ~ "CH",
      near(longitude, -104.7663, tol = 0.0001) ~ "PT",
      near(longitude, -104.9325, tol = 0.0001) ~ "LS",
      near(longitude, -104.8500, tol = 0.0001) ~ "P-30"
    ) %>% as.factor()
  )

# samples remaining after filtering for at least 30% completeness
samples_ind70 <- c("CB0421", "CB0430", "CB0432", "CB0434", "CB0436", "CB0438", "CB0439", "CB0442", "CB0444", "CB0446", "CB0447", "CB0449", "CB0450", "CB0452", "CB0456", "CB0458", "CB0459", "CB0460", "CB0463", "CB0464", "CB0468", "CB0485", "CB0487", "CB0489", "CB0492", "CB0518", "CB0519", "CB0545", "CB0548", "CB0549", "CB0550", "CB0551", "CB0554", "CB0555", "CB0557", "CB0559", "CB0561", "CB0573", "CB0574", "CB0576", "CB0589", "CB0593", "CB0598", "CB0603", "CB0608", "CB0612", "CB0614", "CB0615", "CB0616", "CB0621", "CB0622", "CB0623", "CB0626", "CB0635", "CB0636", "CB0638", "CB0639", "CB0641", "CB0643", "CB0644", "CB0648", "CB0651", "CB0652", "CB0654", "CB0659", "CB0663", "CB0664", "CB0665", "CB0667", "CB0670", "CB0676", "CB0678", "CB0680", "CB0681", "CB0682", "CB0684", "CB0685", "CB0688", "CB0690")

pops_ind70 <- pops %>% 
  filter(individuals %in% samples_ind70)

alt_na_70 <- read_csv(here("output", "genotype_matrices", "genotypes_na_70.csv"))
alt_na_50 <- read_csv(here("output", "genotype_matrices", "genotypes_na_50.csv"))
alt_na_50_ind70 <- read_csv(here("output", "genotype_matrices", "genotypes_na_50_ind70.csv"))

```


Overall missingness. There are individuals missing a ton of data.  

```{r}
miss_df <- alt_na_50_ind70 %>% 
  mutate(pop = pops_ind70$pops,
         site_name = pops_ind70$locality)

miss_mat <- data.matrix(miss_df)

miss_rast <- raster(miss_mat)

plot(miss_rast, col = "black", axes = FALSE)

```

## CH

CH missingness.  

```{r}
miss_df_CH <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "CH")

miss_mat_CH <- data.matrix(miss_df_CH)

miss_rast_CH <- raster(miss_mat_CH)

plot(miss_rast_CH, col = "black", axes = FALSE)
```

## PC

PC missingness.  

```{r}
miss_df_PC <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "PC")

miss_mat_PC <- data.matrix(miss_df_PC)

miss_rast_PC <- raster(miss_mat_PC)

plot(miss_rast_PC, col = "black", axes = FALSE)
```

## LG

LG missingness.  

```{r}
miss_df_LG <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "LG")

miss_mat_LG <- data.matrix(miss_df_LG)

miss_rast_LG <- raster(miss_mat_LG)

plot(miss_rast_LG, col = "black", axes = FALSE)
```

## ST

ST missingness.  

```{r}
miss_df_ST <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "ST")

miss_mat_ST <- data.matrix(miss_df_ST)

miss_rast_ST <- raster(miss_mat_ST)

plot(miss_rast_ST, col = "black", axes = FALSE)
```


## P-35

P-35 missingness.  

```{r}
miss_df_P35 <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "P-35")

miss_mat_P35 <- data.matrix(miss_df_P35)

miss_rast_P35 <- raster(miss_mat_P35)

plot(miss_rast_P35, col = "black", axes = FALSE)
```

## PZ

PZ missingness.  

```{r}
miss_df_PZ <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "PZ")

miss_mat_PZ <- data.matrix(miss_df_PZ)

miss_rast_PZ <- raster(miss_mat_PZ)

plot(miss_rast_PZ, col = "black", axes = FALSE)
```

## PT

PT missingness.  

```{r}
miss_df_PT <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "PT")

miss_mat_PT <- data.matrix(miss_df_PT)

miss_rast_PT <- raster(miss_mat_PT)

plot(miss_rast_PT, col = "black", axes = FALSE)
```


## LS

LS missingness.  

```{r}
miss_df_LS <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "LS")

miss_mat_LS <- data.matrix(miss_df_LS)

miss_rast_LS <- raster(miss_mat_LS)

plot(miss_rast_LS, col = "black", axes = FALSE)
```

## P-30

P-30 missingness.  

```{r}
miss_df_P30 <- alt_na_70 %>% 
  mutate(pop = pops$pops,
         site_name = pops$locality) %>% 
  filter(site_name == "P-30")

miss_mat_P30 <- data.matrix(miss_df_P30)

miss_rast_P30 <- raster(miss_mat_P30)

plot(miss_rast_P30, col = "black", axes = FALSE)
```


Everything together  

```{r}
par(mfrow = c(3,3))
plot(miss_rast_CH, col = "black", axes = FALSE, main = "CH", legend=FALSE)
plot(miss_rast_PC, col = "black", axes = FALSE, main = "PC", legend=FALSE)
plot(miss_rast_LG, col = "black", axes = FALSE, main = "LG", legend=FALSE)
plot(miss_rast_ST, col = "black", axes = FALSE, main = "ST", legend=FALSE)
plot(miss_rast_P35, col = "black", axes = FALSE, main = "P35", legend=FALSE)
plot(miss_rast_PZ, col = "black", axes = FALSE, main = "PZ", legend=FALSE)
plot(miss_rast_PT, col = "black", axes = FALSE, main = "PT", legend=FALSE)
plot(miss_rast_LS, col = "black", axes = FALSE, main = "LS", legend=FALSE)
plot(miss_rast_P30, col = "black", axes = FALSE, main = "P30", legend=FALSE)
```



## No high-missingness

```{r}
miss_df_ind50 <- miss_df[miss_df_50 <= 0.3,]

miss_df_ind50 %>% count(locality)
```

# Session info
```{r}
sessioninfo::session_info()
```
